protocol/binary.go

```diff
@@ -61,3 +67,53 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	e, err := reader.ReadEnveloped()
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
+// specific purpose of decoding requests that may or may not have an envelope.
+// This allows a Thrift request handler to transparently accept requests
+// regardless of whether the caller is configured to submit envelopes.
+func (b binaryProtocol) DecodeRequest(r io.ReaderAt) (wire.Value, Enveloping, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
+	// without a version (non-strict)
+	// 3. A message of length >1 with MSB 0x80 set can only be enveloped with a
+	// version (strict)
+	// 4. A message of length >1 starting with 0x00-0x0f can only be the
+	// beginning of an un-enveloped field.
+	var buf [2]byte
+	// If we fail to read two bytes, the only possible valid value is the empty struct.
+	if _, err := r.ReadAt(buf[0:2], 0); err != nil {
+		val, err := b.Decode(r, wire.TStruct)
+		if err != nil {
+			return wire.Value{}, NoMessage, err
+		}
+		return val, NoEnvelope, nil
+	}
+
+	// 0x00 is a valid preamble for a non-strict enveloped request.
+	if buf[0] == 0x00 {
+		e, err := b.DecodeEnveloped(r)
+		if err != nil {
+			return wire.Value{}, NoMessage, err
+		}
+		return e.Value, EnvelopedMessage, nil
+	}
+
+	// 0x80 is a valid preamble for a strict enveloped request.
+	if buf[0]&0x80 > 0 {
+		e, err := b.DecodeEnveloped(r)
+		if err != nil {
+			return wire.Value{}, NoMessage, err
+		}
+		return e.Value, VersionedMessage, nil
+	}
+
+	// All other patterns are either bare structs or invalid.
```

Comment:
rule 4 was,
```
	// 4. A message of length >1 starting with 0x00-0x0f can only be the
	// beginning of an un-enveloped field.
```

However, we're assuming here that anything where `buf[0] != 0` must be a non-enveloped struct. Isn't it possible it's an enveloped non-strict value with a long string as the name?

If we're ignoring that case, then it seems like we should very explicitly call it out.